### PR TITLE
initial @expect() implementation

### PR DESCRIPTION
**Pull request by shawnl**
Closes:  #489

### CODE REVIEW

**Comment by data-man** [comment]
@andrewrk Frendly ping. :)

File: src/ir.cpp [hunk 1]
```diff
@@ -23420,6 +23452,82 @@ static IrInstruction *ir_analyze_instruction_align_of(IrAnalyze *ira, IrInstruct
...
+                        buf_sprintf("did you reverse the arguments?"));
+                }
+                return ira->codegen->invalid_instruction;
+            }
+        case ZigTypeIdBoundFn:
+        case ZigTypeIdFn:
+        case ZigTypeIdPointer:
+        case ZigTypeIdOptional:
+        case ZigTypeIdErrorUnion:
+            zig_panic("TODO. more types for @expect()");
```
Comment (andrewrk) [inline src/ir.cpp]: `@expect` should work on all types. Instead of panicking here, just return `value`.

File: src/ir.cpp [hunk 2]
```diff
@@ -23420,6 +23452,82 @@ static IrInstruction *ir_analyze_instruction_align_of(IrAnalyze *ira, IrInstruct
...
+    if (type_is_invalid(value->value.type))
+        return ira->codegen->invalid_instruction;
+
+    ZigType *value_type = value->value.type;
+
+    IrInstruction *expect_instr = instruction->expected->child;
+    if (type_is_invalid(expect_instr->value.type))
+        return ira->codegen->invalid_instruction;
+
+    switch (value_type->id) {
```
Comment (andrewrk) [inline src/ir.cpp]: I don't think this switch is needed at all. Implicit cast the expected value to the correct type, resolve it as comptime, and then either return `value` (in the case `value` is comptime known) or return an expect instruction.

**Comment by andrewrk** [comment]
Implementation is not quite right, and it's a low-priority feature (compared to other features in this release cycle). Closing until a more mergeable implementation is available.

**Pull request closed without merging**